core/miner/storage_fsm/impl/commit_batcher_impl.cpp

```diff
@@ -0,0 +1,246 @@
+#include "vm/actor/builtin/v6/monies.hpp"
+
+#include <string>
+
+namespace fc::mining {
+  using api::kPushNoSpec;
+  using fc::BytesIn;
+  using fc::proofs::ProofEngine;
+  using primitives::ActorId;
+  using primitives::go::bigdiv;
+  using primitives::sector::AggregateSealVerifyInfo;
+  using primitives::sector::AggregateSealVerifyProofAndInfos;
+  using primitives::tipset::TipsetCPtr;
+  using vm::actor::MethodParams;
+  using vm::actor::builtin::types::miner::kChainFinality;
+  using vm::actor::builtin::types::miner::SectorPreCommitOnChainInfo;
+  using vm::actor::builtin::v5::miner::ProveCommitAggregate;
+  using vm::actor::builtin::v6::miner::AggregateProveCommitNetworkFee;
+
+  CommitBatcherImpl::CommitBatcherImpl(const std::chrono::seconds &max_time,
+                                       std::shared_ptr<FullNodeApi> api,
+                                       Address miner_address,
+                                       std::shared_ptr<Scheduler> scheduler,
+                                       AddressSelector address_selector,
+                                       std::shared_ptr<FeeConfig> fee_config,
+                                       const size_t &max_size_callback,
+                                       std::shared_ptr<ProofEngine> proof)
+      : scheduler_(std::move(scheduler)),
+        max_delay_(max_time),
+        closest_cutoff_(max_time),
+        max_size_callback_(max_size_callback),
+        api_(std::move(api)),
+        miner_address_(std::move(miner_address)),
+        fee_config_(std::move(fee_config)),
+        proof_(std::move(proof)),
+        address_selector_(std::move(address_selector)) {
+    cutoff_start_ = std::chrono::system_clock::now();
+    reschedule(max_delay_);
+  }
+
+  outcome::result<void> CommitBatcherImpl::addCommit(
+      const SectorInfo &sector_info,
+      const AggregateInput &aggregate_input,
+      const CommitCallback &callback) {
+    std::unique_lock<std::mutex> locker(mutex_storage_);
+
+    const SectorNumber &sector_number = sector_info.sector_number;
+    OUTCOME_TRY(head, api_->ChainHead());
+
+    pair_storage_[sector_number] = PairStorage{aggregate_input, callback};
+
+    if (pair_storage_.size() >= max_size_callback_) {
+      locker.unlock();
+      forceSend();
+      return outcome::success();
+    }
+
+    setCommitCutoff(head->epoch(), sector_info);
+    return outcome::success();
+  }
+
+  void CommitBatcherImpl::forceSend() {
+    MapPairStorage pair_storage_for_send_;
+
+    std::unique_lock<std::mutex> locker(mutex_storage_);
+    pair_storage_for_send_ = std::move(pair_storage_);
+    locker.unlock();
+
+    const auto maybe_result = sendBatch(pair_storage_for_send_);
+
+    for (auto &[key, pair_storage] : pair_storage_for_send_) {
+      pair_storage.commit_callback(maybe_result);
+    }
+
+    cutoff_start_ = std::chrono::system_clock::now();
+    closest_cutoff_ = max_delay_;
+    reschedule(max_delay_);
```

Comment:
maybe better to store queued commits with cutoff, sorted by cutoff, and reschedule by minimal cutoff from queue